trin-storage/src/versioned/id_indexed_v1/pruning_strategy.rs

```diff
@@ -0,0 +1,360 @@
+/// `optimal_pruning_duration_range`
+/// - it will decrease by `max_pruning_count_change_fraction` if pruning duration is above the
+/// `optimal_pruning_duration_range`
+pub struct PruningStrategy {
+    /// The store configuration.
+    config: IdIndexedV1StoreConfig,
+    /// The maximum number of entries to prune at the time.
+    max_pruning_count: u64,
+}
+
+impl PruningStrategy {
+    /// The starting value for `max_pruning_count`.
+    pub const STARTING_MAX_PRUNING_COUNT: u64 = 100;
+
+    pub fn new(config: IdIndexedV1StoreConfig) -> Self {
+        Self {
+            config,
+            max_pruning_count: Self::STARTING_MAX_PRUNING_COUNT,
+        }
+    }
+
+    /// The capacity that we aim for when pruning.
+    pub fn target_capacity_bytes(&self) -> u64 {
+        (self.config.storage_capacity_bytes as f64
+            * self.config.pruning_config.target_capacity_fraction)
+            .round() as u64
+    }
+
+    /// Returns `true` when used capacity is above target capacity.
+    pub fn is_usage_above_target_capacity(&self, usage_stats: &UsageStats) -> bool {
+        usage_stats.is_above(self.target_capacity_bytes())
+    }
+
+    /// Returns `true` when used capacity is above storage capacity.
+    pub fn should_prune(&self, usage_stats: &UsageStats) -> bool {
+        usage_stats.is_above(self.config.storage_capacity_bytes)
+    }
+
+    /// Returns the number of entries to prune.
+    pub fn get_pruning_count(&self, usage_stats: &UsageStats) -> u64 {
+        if !self.should_prune(usage_stats) {
+            return 0;
+        }
+
+        // If storage capacity is 0, prune everything.
+        if self.config.storage_capacity_bytes == 0 {
+            debug!(
+                Db = %self.config.content_type,
+                "Storage capacity is 0. Pruning everything ({})",
+                usage_stats.entry_count
+            );
+            return usage_stats.entry_count;
+        }
+
+        self.estimate_to_delete_until_target(usage_stats)
+            .min(self.max_pruning_count)
+    }
+
+    /// Should be called after pruning in order to update `max_pruning_count` based on pruning
+    /// duration.
+    pub fn observe_pruning_duration(&mut self, pruning_duration: Duration) {
+        let pruning_config = &self.config.pruning_config;
+        let optimal_pruning_duration = &pruning_config.optimal_pruning_duration_range;
+
+        let change_ratio = if pruning_duration < optimal_pruning_duration.start {
+            debug!(
+                Db = %self.config.content_type,
+                "Pruning was too fast. Increasing max_pruning_count",
+            );
+            1. + pruning_config.max_pruning_count_change_fraction
+        } else if pruning_duration > optimal_pruning_duration.end {
+            debug!(
+                Db = %self.config.content_type,
+                "Pruning was too slow. Decreasing max_pruning_count",
+            );
+            1. - pruning_config.max_pruning_count_change_fraction
```

Comment:
Ah, I see now where the upper limit of 1 came from. My intuition before getting to this code is that a change fraction of 1.0 would double on the upside and cut in half on the downside. Those seem like the more symmetric options to me. (ie~ one increase and one decrease takes you back to where you started)

In this approach, choosing a value of .9 nearly doubles on the upside, but is 1/10th on the downside, which is aggressively dropping but not that aggressively climbing (it takes more than 3 upside adjustments to make up for 1 downside adjustment). So if we want one number to govern both directions, my intuition says that this line should be:
```suggestion
            1. / (1. + pruning_config.max_pruning_count_change_fraction)
```